File: Android/melody-runtime/src/main/java/com/melody/runtime/renderer/CustomComponentView.kt
package com.melody.runtime.renderer

import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import com.melody.core.schema.CustomComponentDefinition
import com.melody.core.schema.Value
import com.melody.runtime.engine.LuaValue

/**
 * Renders a custom component template with resolved props.
 * Port of iOS CustomComponentView.swift.
 */
@Composable
fun CustomComponentView(
    template: CustomComponentDefinition,
    instanceProps: Map<String, Value<String>>?
) {
    val luaVM = LocalLuaVM.current
    val parentProps = LocalComponentProps.current

    val resolvedProps = resolveProps(template, instanceProps, luaVM, parentProps)

    CompositionLocalProvider(
        LocalComponentProps provides resolvedProps
    ) {
        ComponentRenderer(components = template.body)
    }
}

private fun resolveProps(
    template: CustomComponentDefinition,
    instanceProps: Map<String, Value<String>>?,
    luaVM: com.melody.runtime.engine.LuaVM?,
    parentProps: Map<String, LuaValue>?
): Map<String, LuaValue> {
    val result = mutableMapOf<String, LuaValue>()

    template.props?.forEach { (key, stateValue) ->
        result[key] = LuaValue.fromStateValue(stateValue)
    }

    if (instanceProps != null && luaVM != null) {
        val prefix = parentProps?.let { luaVM.propsPrefix(it) } ?: ""
        for ((key, propValue) in instanceProps) {
            when (propValue) {
                is Value.Literal -> result[key] = LuaValue.StringVal(propValue.value)
                is Value.Expression -> {
                    try {
                        result[key] = luaVM.execute("${prefix}return ${propValue.expr}")
                    } catch (_: Exception) {
                        result[key] = LuaValue.StringVal(propValue.expr)
                    }
                }
            }
        }
    }

    return result
}

File: Sources/Runtime/Renderer/CustomComponentView.swift
import SwiftUI
import Core

// MARK: - Environment Keys

private struct CustomComponentsKey: EnvironmentKey {
    static let defaultValue: [String: CustomComponentDefinition] = [:]
}

private struct ComponentPropsKey: EnvironmentKey {
    static let defaultValue: [String: LuaValue]? = nil
}

extension EnvironmentValues {
    var customComponents: [String: CustomComponentDefinition] {
        get { self[CustomComponentsKey.self] }
        set { self[CustomComponentsKey.self] = newValue }
    }

    var componentProps: [String: LuaValue]? {
        get { self[ComponentPropsKey.self] }
        set { self[ComponentPropsKey.self] = newValue }
    }
}

// MARK: - Custom Component View

/// Renders a reusable component template with resolved instance props.
struct CustomComponentView: View {
    let template: CustomComponentDefinition
    let instanceProps: [String: Value<String>]?

    @Environment(\.luaVM) private var luaVM
    @Environment(\.componentProps) private var parentProps

    var body: some View {
        let resolvedProps = resolveProps()
        ComponentRenderer(components: template.body)
            .environment(\.componentProps, resolvedProps)
    }

    private func resolveProps() -> [String: LuaValue] {
        var result: [String: LuaValue] = [:]

        if let defaults = template.props {
            for (key, stateValue) in defaults {
                result[key] = luaValue(from: stateValue)
            }
        }

        if let instanceProps, let vm = luaVM {
            let prefix: String
            if let parentProps {
                prefix = vm.localPrefix(key: "props", for: parentProps)
            } else {
                prefix = ""
            }
            for (key, propValue) in instanceProps {
                switch propValue {
                case .literal(let str):
                    result[key] = .string(str)
                case .expression(let expr):
                    if let value = try? vm.execute(prefix + "return " + expr) {
                        result[key] = value
                    } else {
                        result[key] = .string(expr)
                    }
                }
            }
        }

        return result
    }

    private func luaValue(from stateValue: StateValue) -> LuaValue {
        switch stateValue {
        case .string(let s): return .string(s)
        case .int(let i): return .number(Double(i))
        case .double(let d): return .number(d)
        case .bool(let b): return .bool(b)
        case .null: return .nil
        case .array(let arr): return .array(arr.map { luaValue(from: $0) })
        case .dictionary(let dict): return .table(dict.mapValues { luaValue(from: $0) })
        }
    }
}
